internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,332 @@
+	// small batch can lead to poor performance and resource waste in random write workload.
+	// 10K batch size is large enough to get good performance with random write workloads in tests.
+	MinFlushKeys = 10000
+	// MinFlushSize is the minimum size of MemDB to trigger flush.
+	MinFlushSize = 16 * 1024 * 1024 // 16MB
+	// ForceFlushSizeThreshold is the threshold to force flush MemDB, which controls the max memory consumption of PipelinedMemDB.
+	ForceFlushSizeThreshold = 128 * 1024 * 1024 // 128MB
+)
+
+type pipelinedMemDBSkipRemoteBuffer struct{}
+
+// TODO: skip remote buffer by context is too obscure, add a new method to read local buffer.
+var pipelinedMemDBSkipRemoteBufferKey = pipelinedMemDBSkipRemoteBuffer{}
+
+// WithPipelinedMemDBSkipRemoteBuffer is used to skip reading remote buffer for saving RPC.
+func WithPipelinedMemDBSkipRemoteBuffer(ctx context.Context) context.Context {
+	return context.WithValue(ctx, pipelinedMemDBSkipRemoteBufferKey, struct{}{})
+}
+
+type FlushFunc func(uint64, *MemDB) error
+type BufferBatchGetter func(ctx context.Context, keys [][]byte) (map[string][]byte, error)
+
+func NewPipelinedMemDB(bufferBatchGetter BufferBatchGetter, flushFunc FlushFunc) *PipelinedMemDB {
+	return &PipelinedMemDB{
+		memDB:             newMemDB(),
+		errCh:             make(chan error, 1),
+		flushFunc:         flushFunc,
+		bufferBatchGetter: bufferBatchGetter,
+		generation:        0,
+		entryLimit:        math.MaxUint64,
+		bufferLimit:       math.MaxUint64,
+	}
+}
+
+// Dirty returns whether the pipelined buffer is mutated.
+func (p *PipelinedMemDB) Dirty() bool {
+	return p.memDB.Dirty() || p.len > 0
+}
+
+// GetMemDB implements MemBuffer interface.
+func (p *PipelinedMemDB) GetMemDB() *MemDB {
+	panic("GetMemDB should not be invoked for PipelinedMemDB")
+}
+
+// Get the value by given key, it returns tikverr.ErrNotExist if not exist.
+// The priority of the value is MemBuffer > flushingMemDB > flushed memdbs.
+func (p *PipelinedMemDB) Get(ctx context.Context, k []byte) ([]byte, error) {
+	v, err := p.memDB.Get(k)
+	if err == nil {
+		return v, nil
+	}
+	if !tikverr.IsErrNotFound(err) {
+		return nil, err
+	}
+	if p.flushingMemDB != nil {
+		v, err = p.flushingMemDB.Get(k)
+		if err == nil {
+			return v, nil
+		}
+		if !tikverr.IsErrNotFound(err) {
+			return nil, err
+		}
+	}
+	if ctx.Value(pipelinedMemDBSkipRemoteBufferKey) != nil {
+		return nil, tikverr.ErrNotExist
+	}
+	// read remote buffer
+	var (
+		dataMap map[string][]byte
+		ok      bool
+	)
+	dataMap, err = p.bufferBatchGetter(ctx, [][]byte{k})
+	if err != nil {
+		return nil, err
+	}
+	v, ok = dataMap[string(k)]
+	if !ok {
+		return nil, tikverr.ErrNotExist
+	}
+	return v, nil
+}
+
+func (p *PipelinedMemDB) GetFlags(k []byte) (kv.KeyFlags, error) {
+	f, err := p.memDB.GetFlags(k)
+	if p.flushingMemDB != nil && tikverr.IsErrNotFound(err) {
+		f, err = p.flushingMemDB.GetFlags(k)
+	}
+	if err != nil {
+		return 0, err
```

Comment:
This can be replaced by assertion checks in TiKV side when handling flush request.

`GetFlags` should not be called and return an error, let's leave it as future work.